Major structural overhaul to the repo

ryanwelcher opened the pull request:
This PR is an attempt to simplify the structure of the repo by using a single build process for all of the examples. 

Benefits:

- Allows the repo to show off the power and flexibility of the `@wordpress/scripts` package by leveraging both the built-in handling of multiple blocks while also adding our own functionality via an extended `webpack.config.js` file.
- Simplifies the repo by no longer needing to leverage Lerna to build out the various examples.
- Allows all files to be stored in a single `build` directory. This will make using an Action to generate a new .zip archive much easier and potentially allow us to not have to version the `build` directory - which is not a common practice in the wild.

Changes:
- New blocks must be added to the `src` directory as would be done normally when using the scripts package.
- Any non-block examples such as data API examples or slot fill examples must be added to the non-block-examples directory for the build process to find them

ryanwelcher committed:
Update packages to latest.

ryanwelcher committed:
Add start command for blocks

ryanwelcher committed:
Move all blocks to the src directory.

ryanwelcher committed:
Add the --webpack-copy-php flag to ensure all php file are copied to the build directory.

ryanwelcher committed:
Use the same version as Gutenberg.

ryanwelcher committed:
Remove lerna and update dependencies

ryanwelcher committed:
Add custom webpack to allow for one build process for all examples.

ryanwelcher committed:
Re-organize the repo and remove all build processes from each example.

ryanwelcher committed:
Ignore the non-JSX examples for the main linting command.

ryanwelcher committed:
Add the copywebpack plugin as a dependency.

ryanwelcher committed:
Finalize file naming and webpack.config.js set up.

ryanwelcher committed:
Linting fixes.

ryanwelcher committed:
Use a single .eslintrc for all code. Non-JSX examples contain their own for IDE integration.

ryanwelcher committed:
Update readme for developement instructions.

ryanwelcher committed:
Remove trailing comma.

ryanwelcher committed:
Rebuild.

gziolo commented:
> Allows all files to be stored in a single build directory. This will make using an Action to generate a new .zip archive much easier and potentially allow us to not have to version the build directory - which is not a common practice in the wild.

Love the idea. I had the same thoughts when looking at the new folder structure.

> New blocks must be added to the src directory as would be done normally when using the scripts package.

What prevents us from keeping all types of blocks in `src` folder? example:

--| src
----| blocks-plain
----| blocks-jsx
----| other

An alternative would be to make the `src` folder customizable inside `wp-scripts build`.

Do you know why we need `languages` folders for blocks? I don't think it's necessary anymore when you publish your plugin to the Plugin Directory.

gziolo commented on .nvmrc:
```diff
@@ -1 +1 @@
-lts/*
\ No newline at end of file
+14
```
Node 16 is the currently active version. We use 14 in Gutenberg only because we are behind with npm v6 upgrade 😞

ryanwelcher replied:
I've updated this to use `lts/*`.

gziolo commented:
How changes proposed in https://github.com/WordPress/gutenberg-examples/pull/193 relate to this PR?

ryanwelcher committed:
Merge branch 'trunk' into try/single-build-process

ryanwelcher committed:
Update dynamic block to use new structure.

ryanwelcher committed:
Remove and gitigore the build directory.

ryanwelcher committed:
Use most recent node version.

ryanwelcher commented:
> How changes proposed in #193 relate to this PR?

I think that this PR takes precedence over #193 from a code perspective. Once this has landed, I would adjust #193 to introduce the linting actions as well as the build/release ones mentioned here.

ryanwelcher commented:
> What prevents us from keeping all types of blocks in `src` folder? example:

This is tricky, for anything that contains a block.json file, everything works as expected and all the assets are copied correctly to the correct folders.

For the non-block examples, however, only the `index.php` is copied, unless we drop a `block.json` in there which will be confusing from the perspective of someone trying to learn. I would love to not have to use a custom webpack config here but not sure we can accomplish that without changing the `@wordpress/scripts` package to handle non-blocks assets the in the same way as block assets

gziolo commented:
> An alternative would be to make the `src` folder customizable inside `wp-scripts` build.

Do you think we should try renaming `src` to `blocks-jsx` once `wp-scripts` is able to work with custom source folders?

> Do you know why we need `languages` folders for blocks? I don't think it's necessary anymore when you publish your plugin to the Plugin Directory.

This needs to be confirmed.

<img width="350" alt="Screenshot 2022-03-17 at 10 46 35" src="https://user-images.githubusercontent.com/699132/158782444-90578352-a979-4f53-94c8-3ba1a46a0f47.png">

The current folder structure is a bit confusing when you see it for the first time. We need at least to document it because `src` might trick people in both ways:
- they can think it's code for the plugin
- they can think all examples are there

I think it's important to name top-level folders correctly and document them in the README file.

gziolo reviewed (commented):
I like the general direction outline in the README, but it definitely requires more iterations. It would be great to gather feedback from more folks and some detailed testing to ensure all examples still work. I haven't had a chance to test myself yet.

ryanwelcher commented:
> Do you think we should try renaming `src` to `blocks-jsx` once `wp-scripts` is able to work with custom source folders?

That works for me!

I've done a PR for allowing custom src directories in https://github.com/WordPress/gutenberg/pull/39618

metalandcoffee commented:
As I got through testing each block (currently gathering all feedback for a single comment), one thing I wanted to mention is that I don't think the example folders should be prefixed with numbers anymore due to the restructuring.

I also think we should add a README.md to each example folder to give a quick preview of what the block does. I'd be happy to work on this if you'd like. Let me know how I should go about doing so if it's a yes :)

salcode commented:
## Review of this PR

I've tested the following in my local install and each one looks good 👍

- `01-basic`
- `01-basic-esnext`
- `02-stylesheets`
- `03-editable`
- `03-editable-esnext`
- `04-controls`
- `04-controls-esnext`
- `05-recipe-card`
- `05-recipe-card-esnext`
- `06-inner-blocks`
- `06-inner-blocks-esnext`
- `07-slotfills-esnext`
- `08-block-supports-esnext`
- `09-code-data-basics-esnext`
- `10-dynamic-block`
- `format-api`
- `meta-block`
- `plugin-sidebar`

I did notice `npm run test` failed for me but it fails in the same way with `trunk`, so that does not appear to be an issue with this PR.

## Adding README.md files

In response to @metalandcoffee's comment

> I also think we should add a README.md to each example folder to give a quick preview of what the block does. I'd be happy to work on this if you'd like. Let me know how I should go about doing so if it's a yes :)

 I've opened Issue #202 _Adding a `README.md` to each example folder_

## Changing src to blocks-jsx

> Do you think we should try renaming `src` to `blocks-jsx` once `wp-scripts` is able to work with custom source folders?

I would vote against this. I like the idea that this repo represents the "standard approach" and in my experience that is using a directory called `src`.

ryanwelcher commented:
> As I got through testing each block (currently gathering all feedback for a single comment), one thing I wanted to mention is that I don't think the example folders should be prefixed with numbers anymore due to the restructuring.

This makes a lot of sense to me and as the repo grows with examples, the numbering will be less and less relevant

> I also think we should add a README.md to each example folder to give a quick preview of what the block does. I'd be happy to work on this if you'd like. Let me know how I should go about doing so if it's a yes :)

I think this is a great suggestion!

At a high level, we should have the following in each README

1. Overview of what the block does
2. Concepts being demonstrated
3. Any relevant links to docs
4. Does a screenshot make sense?

@metalandcoffee if you want to get a PR together for the first one, that'd be fantastic!

ryanwelcher commented:
There is an existing issue for adding README files - https://github.com/WordPress/gutenberg-examples/issues/45, lets track the work against that

ryanwelcher committed:
Merge branch 'trunk' into try/single-build-process

ryanwelcher committed:
Regenerate package-lock.json after merge.

ryanwelcher committed:
Format JS files.

ryanwelcher committed:
Merge branch 'trunk' into try/single-build-process

ryanwelcher committed:
Rename src directory to blocks-jsx.

ryanwelcher committed:
Rename non-JSX-examples directory to blocks-non-jsx.

ryanwelcher committed:
Update packcages and use --webpack-src-dir flag to point to blocks-jsx instead of src.

ryanwelcher committed:
Use nonBlockExamplesDir instead of otherFolder to better reflect the purpose of the variable.

gziolo commented on .github/workflows/nodejs.yml:
```diff
@@ -3,25 +3,24 @@ name: Node CI
-    runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                node-version: [10.x, 12.x, 14.x]
```
I recommend using Node.js 14, 16 and 18 - all current LTE versions.

ryanwelcher replied:
This action is going to be changed but I'll update this now - thanks!

gziolo commented on non-block-examples/format-api/index.js:
```diff
@@ -0,0 +1,30 @@
+import { registerFormatType, toggleFormat } from '@wordpress/rich-text';
+import { RichTextToolbarButton } from '@wordpress/block-editor';
+
+const MyCustomButton = ({ isActive, onChange, value }) => {
```
Code formatting looks like a regular Prettier instead of the `wp-prettier` fork 😅

ryanwelcher replied:
🤦 I can't even tell the difference any more 😆

ryanwelcher replied:
Locally I added `prettier": "npm:wp-prettier@2.2.1-beta-1",` and ran `format` from `wp-scripts` which seemed to work so I pushed those changes update. I'm in a weird state where I have my eslint wanting one thing and format another so I'll have to get that sorted out.

ryanwelcher committed:
Create two separate build processes for blocks vs non-block examples.

ryanwelcher committed:
Create separate build process.

ryanwelcher committed:
Merge branch 'try/single-build-process' of github.com:WordPress/gutenberg-examples into try/single-build-process

ryanwelcher committed:
Revert "Create two separate build processes for blocks vs non-block examples."

This reverts commit 9ac0fe9f8a8b07608ac7a7a5ef8da6503465386b.

ryanwelcher committed:
Update node versions in action.

ryanwelcher committed:
Revert "Revert "Create two separate build processes for blocks vs non-block examples.""

This reverts commit d703499ef99b9c3de0be7b097c342cc423873915.

ryanwelcher committed:
Remove built files added incorrectly to repo.

ryanwelcher committed:
Run formatting on JS files.

ryanwelcher committed:
Add wp-prettier alias to address isses with the fork not being installed.

ryanwelcher committed:
Run npm install instead of ci

ryanwelcher commented:
I've made a few changes to the approach for this PR:
1. I've created separate build processes for the blocks and non-block examples. This seemed to make the most sense as I needed to add a custom webpack config for the non-blocks.
2. I've renamed `src` to `blocks-jsx` and used the new `--webpack-src-dir` to point to the new directory and changed the non-jsx versions to be `block-non-jsx`
3. I've added the wp-prettier alias as a dependency because of issues with the actually prettier package being installed and causing issues with formatting

ryanwelcher committed:
Adding build commands.

gziolo reviewed (approved):
Let's give it a try and iterate further with follow-up work. It should help with repository management. Awsome work @ryanwelcher.
